tsl/src/fdw/deparse.c

```diff
@@ -867,6 +867,59 @@ deparseDistinctClause(StringInfo buf, deparse_expr_cxt *context)
 			return;
 	}
 
+	if (query->hasDistinctOn)
+	{
+        /*
+         * Pushing down DISTINCT ON is more complex than plain DISTINCT.
+         * The former requires a matching ORDER BY clause, or else the result is
+         * undefined. There is a diagnostic for the case where the ORDER BY
+         * clause doesn't match the DISTINCT ON clause, so in this case we would
+         * get an error on the data node. There is no diagnostic for the case
+         * where the ORDER BY is absent, so in this case we would get a wrong
+         * result.
```

Comment:
If the ORDER BY is absent, can't we just add one? The ORDER BY should not matter for anything else, unless I am missing something.

Reply:
I think so, if we have a DISTINCT ON, we'll have to sort accordingly anyway, and better do it on the data nodes. This is a further optimization that should be done somewhere at the planning stage, not at deparsing. This check I added just restores correctness, preventing us from generating an obviously wrong query. We might decide to switch it to an assertion later.